plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java

```diff
@@ -233,9 +278,94 @@ private ConnectorTableMetadata getTableMetadata(ConnectorSession session, Schema
+                .intersect(handle.getCompactEffectivePredicate())
+                .intersect(withColumnDomains(pushedDown.buildOrThrow()));
+
+        // Get list of all columns involved in predicate
+        Set<String> predicateColumnNames = new HashSet<>();
+        newEffectivePredicate.getDomains().get().keySet().stream()
+                .map(KafkaColumnHandle::getName)
+                .forEach(predicateColumnNames::add);
+
+        boolean isRightForPush = false;
+        if (KafkaSessionProperties.isPredicatePushDownEnabled(session)) {
+            isRightForPush = checkIfPossibleToPush(newEffectivePredicate.getDomains().get().keySet());
+        }
+        // Get the column handle
+        Map<String, ColumnHandle> columnHandles = getColumnHandles(session, table);
+
+        //map predicate columns to kafka column handles
+        Map<String, KafkaColumnHandle> predicateColumns = predicateColumnNames.stream()
+                .map(columnHandles::get)
+                .map(KafkaColumnHandle.class::cast)
+                .collect(toImmutableMap(KafkaColumnHandle::getName, identity()));
+
+        newHandle = new KafkaTableHandle(
+                newHandle.getSchemaName(),
+                newHandle.getTableName(),
+                newHandle.getTopicName(),
+                newHandle.getKeyDataFormat(),
+                newHandle.getMessageDataFormat(),
+                newHandle.getKeyDataSchemaLocation(),
+                newHandle.getMessageDataSchemaLocation(),
+                newHandle.getKeySubject(),
+                newHandle.getMessageSubject(),
+                newHandle.getColumns(),
+                compactEffectivePredicate,
+                effectivePredicate,
+                predicateColumns,
+                isRightForPush,
+                newHandle.getLimit());
+
+        if (handle.getCompactEffectivePredicate().equals(newHandle.getCompactEffectivePredicate())) {
+            return Optional.empty();
+        }
+
+        boolean isFullPushDown;
+        if (isRightForPush) {
+            // Filter some unnecessary columns
+            TupleDomain<ColumnHandle> remainingFilter = newHandle.getRemainingFilter(kafkaInternalFieldManager);
+            isFullPushDown = remainingFilter.isAll();
+            // All predicates might be pushed down at this time, except for the special columns that don't support at right now.
+            // e.g. _message*
+            // discrete range matching for a query like: where x in (1, 6)
+            return Optional.of(new ConstraintApplicationResult<>(newHandle, isFullPushDown ? all() : remainingFilter, false));
+        }
+
+        return Optional.of(new ConstraintApplicationResult<>(handle, effectivePredicate, false));
+    }
+
+    @Override
+    public Optional<LimitApplicationResult<ConnectorTableHandle>> applyLimit(ConnectorSession session, ConnectorTableHandle table, long limit)
```

Comment:
Please avoid mixing up changes for predicate pushdown and limit pushdown. I would recommend separating into some PRs.

Reply:
Predicate pushdown and limit pushdown have a dependency relationship, where limit pushdown relies on the optimization of the AST (if filterNode exists) by predicate pushdown. Therefore, they were merged into a single PR submission. BTW, Could you provide some suggestions on how to split the PR, and how many PRs?